resources/puppetlabs/puppetdb/pql/pql-grammar-experimental.ebnf

```diff
@@ -0,0 +1,139 @@
+
+(* Static list of entity types *)
+<entity> = 'facts' |
+           'reports' |
+           'catalogs' |
+           'resources' |
+           'nodes' |
+           'environments' |
+           'factsets' |
+           'fact_paths' |
+           'fact_contents' |
+           'edges' |
+           'events';
+
+(* Expressions
+
+   This structure preserves precedence rules after parsing, so for reference the number
+   after expr is the order of precedence. One should be careful when modifying
+   this structure, and ensure that precedence rules continue to be kept. *)
+<expression> = expr4;
+expr4   = expr3 { <whitespace>, or, <whitespace>, expr4 };
+expr3   = expr2 { <whitespace>, and, <whitespace>, expr3 };
+expr2   = ( not, [<whitespace>], expr2 ) | expr1;
+<expr1> = ( <lparens>, expression, <rparens> ) | condexpression | subquery;
+
+(* Implicit subqueries *)
+subquery = entity, [<whitespace>], where;
+
+(* Conditional expression *)
+condexpression    = (condexpregexp | condexpnumber | condexpmatch | condexpsubquery | condexpnull);
+<condexpregexp>   = field, [<whitespace>], condregexp, [<whitespace>], valueregexp;
+<condexpnumber>   = field, [<whitespace>], condnumber, [<whitespace>], valuenumber;
+<condexpmatch>    = field, [<whitespace>], condmatch, [<whitespace>], valuematch;
+<condexpsubquery> = (field | groupedfieldlist), [<whitespace>], condsubquery, [<whitespace>], query;
+<condexpnull>     = field, [<whitespace>], condnull, [<whitespace>], boolean;
+
+(* Conditional expression parts *)
+groupedfieldlist = <lparens>, [<whitespace>], fieldlist, [<whitespace>], <rparens>;
+<fieldlist>      = field, [ [<whitespace>], <','>, [<whitespace>], fieldlist ];
+<field>          = #'[a-zA-Z_]+';
+
+<condregexp>    = '~';
+<condnumber>    = '>=' | '<=' | '<' | '>';
+<condmatch>     = '=';
+<condsubquery>  = 'in';
+<condnull>      = 'null?';
+
+<valueregexp>   = regexp;
+<valuenumber>   = integer | real;
+<valuematch>    = string | integer | real | boolean;
+
+(* Boolean operators *)
+<and> = <'and'>;
+<or>  = <'or'>;
+not   = <'!'>;
+
+(* Regexp strings *)
+regexp = <regexpquote>, stringwithoutregexpquote, <regexpquote>
+<stringwithoutregexpquote> = #'(?:[^/\\]|\\.)*';
+<regexpquote> = "/"
+
+(* Basic strings
+
+   Here we differentiate between double quoted and single quoted
+   strings so we can determine how to strip escapes later on. *)
+<string> = dqstring | sqstring;
+dqstring = (<doublequote>, stringwithoutdoublequotes, <doublequote>);
+<stringwithoutdoublequotes> = #'(?:[^"\\]|\\.)*';
+<doublequote> = "\"";
+
+sqstring = (<singlequote>, stringwithoutsinglequotes, <singlequote>);
+<stringwithoutsinglequotes> = #'(?:[^\'\\]|\\.)*';
+<singlequote> = "'";
+
+(* Parens *)
+<lparens> = '(';
+<rparens> = ')';
+
+(* Braces *)
+<lbrace> = '{';
+<rbrace> = '}';
+
+(* Bracket *)
+<lbracket> = '[';
+<rbracket> = ']';
+
+(* Booleans *)
+boolean = true | false;
+true = <"true">;
+false = <"false">;
+
+(* Numbers *)
+integer    = [negative], digits;
```

Comment:
Oh I thought I had done integers, let me fix. Indeed I'll look into the other issue also.

Still, aren't you chuffed I added e-notation support? :-).